packages/access-control/src/lib.rs

```diff
@@ -1,147 +1,123 @@
+use error::{Error, Result};
 use sdk::{
-    cosmwasm_std::{Addr, StdError, StdResult, Storage},
+    cosmwasm_std::{Addr, Storage},
     cw_storage_plus::Item,
 };
-
-const CONTRACT_OWNER_NAMESPACE: &str = "contract_owner";
-
-pub struct SingleUserAccess<'r> {
-    storage_namespace: &'r str,
-    address: Addr,
+use std::ops::{Deref, DerefMut};
+
+mod contract_owner;
+pub use contract_owner::ContractOwnerAccess;
+pub mod error;
+
+pub struct SingleUserAccess<'storage, 'namespace, S>
+where
+    S: Deref<Target = dyn Storage + 'storage>,
+{
+    storage: S,
+    storage_item: Item<'namespace, Addr>,
 }
 
-impl<'r> SingleUserAccess<'r> {
-    pub const fn new(storage_namespace: &'r str, address: Addr) -> Self {
+impl<'storage, 'namespace, S> SingleUserAccess<'storage, 'namespace, S>
+where
+    S: Deref<Target = dyn Storage + 'storage>,
+{
+    pub const fn new(storage: S, storage_namespace: &'namespace str) -> Self {
         Self {
-            storage_namespace,
-            address,
+            storage,
+            storage_item: Item::new(storage_namespace),
         }
     }
 
-    pub fn load(storage: &dyn Storage, storage_namespace: &'r str) -> StdResult<Self> {
-        Item::new(storage_namespace)
-            .load(storage)
-            .map(|address| Self {
-                storage_namespace,
-                address,
-            })
-    }
-
-    pub fn store(&self, storage: &mut dyn Storage) -> StdResult<()> {
-        Item::new(self.storage_namespace).save(storage, &self.address)
+    pub fn check(&self, user: &Addr) -> Result {
+        self.storage_item
+            .load(self.storage.deref())
+            .map_err(Into::into)
+            .and_then(|granted_to| UserPermission::new(granted_to).check(user))
     }
+}
 
-    pub fn remove(storage: &mut dyn Storage, storage_namespace: &'r str) {
-        Item::<()>::new(storage_namespace).remove(storage)
+impl<'storage, 'namespace, S> SingleUserAccess<'storage, 'namespace, S>
+where
+    S: Deref<Target = dyn Storage + 'storage> + DerefMut,
+{
+    pub fn grant_to(&mut self, user: &Addr) -> Result {
+        self.storage_item
+            .save(self.storage.deref_mut(), user)
+            .map_err(Into::into)
     }
+}
 
-    pub const fn address(&self) -> &Addr {
-        &self.address
-    }
+#[derive(Debug, Eq, PartialEq)]
+struct UserPermission {
```

Comment:
I feel like a function/static method would've done the job just fine. This feels a bit like abusing OOP.

Reply:
true, did not expect initially that the impl would be so simple